gemini/function-calling/sql-talk-app/app.py

```diff
@@ -0,0 +1,257 @@
+import time
+import streamlit as st
+from google.cloud import bigquery
+from vertexai.preview.generative_models import (
+    FunctionDeclaration,
+    GenerativeModel,
+    Part,
+    Tool,
+)
+
+sql_query_func = FunctionDeclaration(
+    name="sql_query",
+    description="Get information from data in BigQuery using SQL queries",
+    parameters={
+        "type": "object",
+        "properties": {
+            "query": {
+                "type": "string",
+                "description": "SQL query on a single line that will help give quantitative answers to the user's question when run on a BigQuery dataset and table. In the SQL query, always use the fully qualified dataset and table names.",
+            }
+        },
+        "required": [
+            "query",
+        ],
+    },
+)
+
+list_datasets_func = FunctionDeclaration(
+    name="list_datasets",
+    description="Get a list of datasets that will help answer the user's question",
+    parameters={
+        "type": "object",
+        "properties": {},
+    },
+)
+
+list_tables_func = FunctionDeclaration(
+    name="list_tables",
+    description="List tables in a dataset that will help answer the user's question",
+    parameters={
+        "type": "object",
+        "properties": {
+            "dataset_id": {
+                "type": "string",
+                "description": "Fully qualified ID of the dataset to fetch tables from. Always use the fully qualified dataset and table names.",
+            }
+        },
+        "required": [
+            "dataset_id",
+        ],
+    },
+)
+
+get_table_func = FunctionDeclaration(
+    name="get_table",
+    description="Get information about a table, including the description, schema, and number of rows that will help answer the user's question. Always use the fully qualified dataset and table names.",
+    parameters={
+        "type": "object",
+        "properties": {
+            "table_id": {
+                "type": "string",
+                "description": "Fully qualified ID of the table to get information about",
+            }
+        },
+        "required": [
+            "query",
+        ],
+    },
+)
+
+sql_query_tool = Tool(
+    function_declarations=[
+        sql_query_func,
+        list_datasets_func,
+        list_tables_func,
+        get_table_func,
+    ],
+)
+
+model = GenerativeModel(
+    "gemini-pro", generation_config={"temperature": 0}, tools=[sql_query_tool]
```

Comment:
nit: Can you use the `GenerationConfig` class instead of a dict?

Reply:
Thanks for the note! I'm going to leave it as is since it matches the signatures/approach that's in the [function calling docs](https://cloud.google.com/vertex-ai/docs/generative-ai/multimodal/function-calling). But as we iterate on the Gemini API and code samples, I will keep these two changes in mind.